python/lsst/daf/butler/cli/utils.py

```diff
@@ -440,3 +567,259 @@ def make_metavar(self):
         return metavar
+
+
+class OptionSection(MWOption):
+    """Implements an Option that prints a section label in the help text and
+    does not pass any value to the command function.
+
+    This class does a bit of hackery to add a section label to a click command
+    help output: first, `expose_value` is set to `False` so that no value is
+    passed to the command function. Second, this class overrides
+    `click.Option.get_help_record` to return the section label string without
+    any prefix so that it stands out as a section label.
+
+    The intention for this implementation is to do minimally invasive overrides
+    of the click classes so as to be robust and easy to fix if the click
+    internals change.
+
+    Parameters
+    ----------
+    sectionName : `str`
+        The parameter declaration for this option. It is not shown to the user,
+        it must be unique within the command. If using the `section` decorator
+        to add a section to a command's options, the section name is
+        auto-generated.
+    sectionText : `str`
+        The text to print in the section identifier.
+    """
+
+    def __init__(self, sectionName, sectionText):
+        super().__init__(sectionName, expose_value=False)
+        self.sectionText = sectionText
+
+    def get_help_record(self, ctx):
+        return (self.sectionText, "")
+
+
+class MWOptionDecorator(abc.ABC):
+
+    _name = None
+    _opts = None
+
+    def __init__(self, forward=False):
+        self.forward = forward
+
+    @staticmethod
+    @abc.abstractmethod
+    def defaultHelp():
+        """The help default help text to use for the Option.
+
+        Returns
+        -------
+        defaultHelp : `str`
+            The default help text.
+        """
+        pass
+
+    @staticmethod
+    @abc.abstractmethod
+    def optionFlags():
+        """The flags that become the param_decls of a click.Parameter
+
+        Returns
+        -------
+        flags : `tuple` [`str`]
+            The flags to use.
+        """
+        pass
+
+    @classmethod
+    def _update(cls):
+        """Use the click.Option to generate the kwarg name and the list of
+        flags
+        """
+        if cls._name is None:
+            p = click.Option(cls.optionFlags())
+            cls._name = p.name
+            cls._opts = p.opts
+
+    @classmethod
+    def optionKey(cls):
+        """Get the keyword argument name for an option flag."""
+        cls._update()
+        return cls._name
+
+    @classmethod
+    def flag(cls):
+        """Get the option flag. Prefers the first flag that begins with two
+        dashes if one is declared, otherwise uses the first flag that begins
+        with a single dash."""
+        cls._update()
+        ret = None
+        for opt in cls._opts:
+            if ret is None or (ret.startswith("-") and opt.startswith("--")):
+                ret = opt
+            if ret.startswith("--"):
+                break
+        return ret
```

Comment:
I had to read it twice to figure out what it does, here is my version which I think is more transparent:

    ret = None
    for opt in cls._opts:
        if ret.startswith("--"):
            return opt
        elif ret is None:  # maybe also needs `and opt.stratswith("-")`
            ret = opt
    return ret